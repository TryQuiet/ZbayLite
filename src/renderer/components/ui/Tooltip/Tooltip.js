import React, { useState } from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import * as R from 'ramda'

import MuiTooltip from '@material-ui/core/Tooltip'
import { withStyles } from '@material-ui/core/styles'

function arrowGenerator (color, theme) {
  return {
    zIndex: 10,
    opacity: 1,
    '&[x-placement*="bottom"] $arrow': {
      opacity: 1,
      left: 0,
      marginTop: '-0.95em',

      '&::before': {
        borderWidth: '0 0.5em 0.5em 0.5em',
        borderColor: `transparent transparent ${color} transparent`
      }
    },
    '&[x-placement*="bottom-end"] $arrow': {
      opacity: 1,
      left: `calc(100% - ${theme.spacing(1)}px - ${constants.arrowSize}) !important`,
      marginTop: '-0.95em',
      '&::before': {
        borderWidth: '0 0.5em 0.5em 0.5em',
        borderColor: `transparent transparent ${color} transparent`
      }
    },
    '&[x-placement*="bottom-start"] $arrow': {
      opacity: 1,
      left: `${theme.spacing(1)}px !important`,
      marginTop: '-0.95em',
      '&::before': {
        borderWidth: '0 0.5em 0.5em 0.5em',
        borderColor: `transparent transparent ${color} transparent`
      }
    },
    '&[x-placement*="top"] $arrow': {
      top: 39,
      left: 0,
      marginBottom: '-0.95em',
      '&::before': {
        borderWidth: '0.5em 0.5em 0 0.5em',
        borderColor: `${color} transparent transparent transparent`
      }
    },
    '&[x-placement*="right"] $arrow': {
      left: 0,
      marginLeft: '-0.95em',
      '&::before': {
        borderWidth: '0.5em 0.5em 0.5em 0',
        borderColor: `transparent ${color} transparent transparent`
      }
    },
    '&[x-placement*="left"] $arrow': {
      right: 0,
      marginRight: '-0.95em',
      '&::before': {
        borderWidth: '0.5em 0 0.5em 0.5em',
        borderColor: `transparent transparent transparent ${color}`
      }
    }
  }
}

const constants = {
  arrowSize: '3em'
}

const styles = theme => ({
  noWrap: {
    maxWidth: 'none',
    filter: 'drop-shadow(0 0 0px #aaaaaa)'
  },
  tooltip: {
    marginBottom: 5,
    background: theme.palette.colors.trueBlack,
    color: theme.typography.body1.color,
    paddingTop: 12,
    paddingBottom: 12,
    paddingLeft: 16,
    paddingRight: 16,
    borderRadius: 8
  },
  arrow: {
    position: 'absolute',
    width: constants.arrowSize,
    height: constants.arrowSize,
    top: '0.5em',
    '&::before': {
      content: '""',
      margin: 'auto',
      display: 'block',
      width: 0,
      height: 0,
      borderStyle: 'solid'
    }
  },
  text: {
    color: theme.palette.colors.white,
    fontSize: 12,
    fontWeight: 500
  },
  arrowPopper: arrowGenerator(theme.palette.colors.trueBlack, theme)
})

export const Tooltip = ({ classes, children, title, titleHTML, noWrap, className, onClick, ...props }) => {
  const [arrowRef, setArrowRef] = useState(null)
  return (
    <span onClick={onClick}>
      <MuiTooltip
        {...props}
        title={
          <React.Fragment>
            {titleHTML || <span className={classes.text}>{title.charAt(0).toUpperCase() + title.slice(1)}</span>}
            <span className={classes.arrow} ref={setArrowRef} />
          </React.Fragment>
        }
        classes={{
          tooltip: classNames({
            [classes.noWrap]: noWrap,
            [className]: className,
            [classes.tooltip]: true
          }),
          popper: classes.arrowPopper
        }}
        PopperProps={{
          popperOptions: {
            modifiers: {
              arrow: {
                enabled: Boolean(arrowRef),
                element: arrowRef
              }
            }
          }
        }}
      >
        {children}
      </MuiTooltip>
    </span>
  )
}

const joiningProd = R.compose(
  R.map(R.join('')),
  R.xprod
)

Tooltip.propTypes = {
  classes: PropTypes.object.isRequired,
  children: PropTypes.element.isRequired,
  title: PropTypes.oneOfType([PropTypes.string, PropTypes.object]).isRequired,
  placement: PropTypes.oneOf(joiningProd(['bottom', 'top'], ['-start', '-end', ''])),
  interactive: PropTypes.bool,
  className: PropTypes.string,
  noWrap: PropTypes.bool,
  onClick: PropTypes.func
}

Tooltip.defaultProps = {
  noWrap: true,
  interactive: false,
  className: '',
  placement: 'bottom',
  onClick: () => {}
}

export default R.compose(
  React.memo,
  withStyles(styles)
)(Tooltip)
