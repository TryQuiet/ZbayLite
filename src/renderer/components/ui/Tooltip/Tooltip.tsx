import React, { ReactElement, ReactFragment, useState } from 'react'
import classNames from 'classnames'

import MuiTooltip from '@material-ui/core/Tooltip'
import { makeStyles, Theme } from '@material-ui/core/styles'

function arrowGenerator(color: string, theme: Theme) {
  return {
    zIndex: 10,
    opacity: 1,
    '&[x-placement*="bottom"] $arrow': {
      opacity: 1,
      left: 0,
      marginTop: '-0.95em',

      '&::before': {
        borderWidth: '0 0.5em 0.5em 0.5em',
        borderColor: `transparent transparent ${color} transparent`
      }
    },
    '&[x-placement*="bottom-end"] $arrow': {
      opacity: 1,
      left: `calc(100% - ${theme.spacing(1)}px - ${constants.arrowSize}) !important`,
      marginTop: '-0.95em',
      '&::before': {
        borderWidth: '0 0.5em 0.5em 0.5em',
        borderColor: `transparent transparent ${color} transparent`
      }
    },
    '&[x-placement*="bottom-start"] $arrow': {
      opacity: 1,
      left: `${theme.spacing(1)}px !important`,
      marginTop: '-0.95em',
      '&::before': {
        borderWidth: '0 0.5em 0.5em 0.5em',
        borderColor: `transparent transparent ${color} transparent`
      }
    },
    '&[x-placement*="top"] $arrow': {
      top: 39,
      left: 0,
      marginBottom: '-0.95em',
      '&::before': {
        borderWidth: '0.5em 0.5em 0 0.5em',
        borderColor: `${color} transparent transparent transparent`
      }
    },
    '&[x-placement*="right"] $arrow': {
      left: 0,
      marginLeft: '-0.95em',
      '&::before': {
        borderWidth: '0.5em 0.5em 0.5em 0',
        borderColor: `transparent ${color} transparent transparent`
      }
    },
    '&[x-placement*="left"] $arrow': {
      right: 0,
      marginRight: '-0.95em',
      '&::before': {
        borderWidth: '0.5em 0 0.5em 0.5em',
        borderColor: `transparent transparent transparent ${color}`
      }
    }
  }
}

const constants = {
  arrowSize: '3em'
}

const usestyles = makeStyles(theme => ({
  noWrap: {
    maxWidth: 'none',
    filter: 'drop-shadow(0 0 0px #aaaaaa)'
  },
  tooltip: {
    marginBottom: 5,
    background: theme.palette.colors.trueBlack,
    color: theme.typography.body1.color,
    paddingTop: 12,
    paddingBottom: 12,
    paddingLeft: 16,
    paddingRight: 16,
    borderRadius: 8
  },
  arrow: {
    position: 'absolute',
    width: constants.arrowSize,
    height: constants.arrowSize,
    top: '0.5em',
    '&::before': {
      content: '""',
      margin: 'auto',
      display: 'block',
      width: 0,
      height: 0,
      borderStyle: 'solid'
    }
  },
  text: {
    color: theme.palette.colors.white,
    fontSize: 12,
    fontWeight: 500
  },
  arrowPopper: arrowGenerator(theme.palette.colors.trueBlack, theme)
}))

interface TooltipProps {
  children: ReactElement
  title?: string
  titleHTML?: ReactFragment
  noWrap?: boolean
  interactive?: boolean
  className?: string
  placement?: 'bottom' | 'top' | 'bottom-start' | 'bottom-end' | 'top-start' | 'top-end'
  onClick?: (e: React.MouseEvent) => void
}

export const Tooltip: React.FC<React.ComponentProps<typeof MuiTooltip> & TooltipProps> = ({
  children,
  title,
  titleHTML,
  noWrap = false,
  interactive = false,
  className = '',
  placement = 'bottom',
  onClick = () => { },
  ...props
}) => {
  const classes = usestyles({})
  const [arrowRef, setArrowRef] = useState(null)
  return (
    <span onClick={e => onClick(e)}>
      <MuiTooltip
        {...props}
        title={
          <React.Fragment>
            {titleHTML || (
              <span className={classes.text}>
                {title ? title.charAt(0).toUpperCase() : ''}
                {title ? title.slice(1) : ''}
              </span>
            )}
            <span className={classes.arrow} ref={setArrowRef} />
          </React.Fragment>
        }
        classes={{
          tooltip: classNames({
            [classes.noWrap]: noWrap,
            [className]: className,
            [classes.tooltip]: true
          }),
          popper: classes.arrowPopper
        }}
        PopperProps={{
          popperOptions: {
            modifiers: {
              arrow: {
                enabled: Boolean(arrowRef),
                element: arrowRef
              }
            }
          }
        }}>
        {children}
      </MuiTooltip>
    </span>
  )
}

export default Tooltip
