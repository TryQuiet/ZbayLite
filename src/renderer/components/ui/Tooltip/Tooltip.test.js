/* eslint import/first: 0 */
import React from 'react'
import { shallow } from 'enzyme'

import each from 'jest-each'
import { mockClasses } from '../../../../shared/testing/mocks'
import { Tooltip } from './Tooltip'

describe('Tooltip', () => {
  const TooltipContent = () => <div>TooltipContent</div>
  /* Skipping test because it can't be verified correctly without wrapping component (MuiThemeProvider),
     which cannot be applied using enzyme library - waiting for @testing-library/react to be introduced */
  it.skip('renders component', () => {
    const result = shallow(
      <Tooltip
        classes={mockClasses}
        title='test-title'
      >
        <TooltipContent />
      </Tooltip>
    )
    expect(result).toMatchSnapshot()
  })

  // it('renders component with optional props', () => {
  //   const result = shallow(
  //     <Tooltip
  //       classes={mockClasses}
  //       title='test-title'
  //       interactive
  //       noWrap
  //     >
  //       <TooltipContent />
  //     </Tooltip>
  //   )
  //   expect(result).toMatchSnapshot()
  // })

  // each(['bottom-start', 'bottom', 'bottom-end', 'top-start', 'top', 'top-end']).test(
  //   'renders component with correct arrow placement - %s',
  //   (placement) => {
  //     const result = shallow(
  //       <Tooltip
  //         classes={mockClasses}
  //         title='test-title'
  //         placement={placement}
  //       >
  //         <TooltipContent />
  //       </Tooltip>
  //     )
  //     expect(result).toMatchSnapshot()
  //   }
  // )
})
