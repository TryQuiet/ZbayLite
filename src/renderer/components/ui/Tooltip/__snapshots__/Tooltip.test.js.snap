// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip renders component 1`] = `
<span
  onClick={[Function]}
>
  <WithStyles(ForwardRef(Tooltip))
    PopperProps={
      Object {
        "popperOptions": Object {
          "modifiers": Object {
            "arrow": Object {
              "element": null,
              "enabled": false,
            },
          },
        },
      }
    }
    classes={
      Object {
        "popper": "arrowPopper-mock",
        "tooltip": "noWrap-mock tooltip-mock",
      }
    }
    interactive={false}
    placement="bottom"
    title={
      <React.Fragment>
        <span
          className="text-mock"
        >
          Test-title
        </span>
        <span
          className="arrow-mock"
        />
      </React.Fragment>
    }
  >
    <TooltipContent />
  </WithStyles(ForwardRef(Tooltip))>
</span>
`;

exports[`Tooltip renders component with correct arrow placement - bottom 1`] = `
<span
  onClick={[Function]}
>
  <WithStyles(ForwardRef(Tooltip))
    PopperProps={
      Object {
        "popperOptions": Object {
          "modifiers": Object {
            "arrow": Object {
              "element": null,
              "enabled": false,
            },
          },
        },
      }
    }
    classes={
      Object {
        "popper": "arrowPopper-mock",
        "tooltip": "noWrap-mock tooltip-mock",
      }
    }
    interactive={false}
    placement="bottom"
    title={
      <React.Fragment>
        <span
          className="text-mock"
        >
          Test-title
        </span>
        <span
          className="arrow-mock"
        />
      </React.Fragment>
    }
  >
    <TooltipContent />
  </WithStyles(ForwardRef(Tooltip))>
</span>
`;

exports[`Tooltip renders component with correct arrow placement - bottom-end 1`] = `
<span
  onClick={[Function]}
>
  <WithStyles(ForwardRef(Tooltip))
    PopperProps={
      Object {
        "popperOptions": Object {
          "modifiers": Object {
            "arrow": Object {
              "element": null,
              "enabled": false,
            },
          },
        },
      }
    }
    classes={
      Object {
        "popper": "arrowPopper-mock",
        "tooltip": "noWrap-mock tooltip-mock",
      }
    }
    interactive={false}
    placement="bottom-end"
    title={
      <React.Fragment>
        <span
          className="text-mock"
        >
          Test-title
        </span>
        <span
          className="arrow-mock"
        />
      </React.Fragment>
    }
  >
    <TooltipContent />
  </WithStyles(ForwardRef(Tooltip))>
</span>
`;

exports[`Tooltip renders component with correct arrow placement - bottom-start 1`] = `
<span
  onClick={[Function]}
>
  <WithStyles(ForwardRef(Tooltip))
    PopperProps={
      Object {
        "popperOptions": Object {
          "modifiers": Object {
            "arrow": Object {
              "element": null,
              "enabled": false,
            },
          },
        },
      }
    }
    classes={
      Object {
        "popper": "arrowPopper-mock",
        "tooltip": "noWrap-mock tooltip-mock",
      }
    }
    interactive={false}
    placement="bottom-start"
    title={
      <React.Fragment>
        <span
          className="text-mock"
        >
          Test-title
        </span>
        <span
          className="arrow-mock"
        />
      </React.Fragment>
    }
  >
    <TooltipContent />
  </WithStyles(ForwardRef(Tooltip))>
</span>
`;

exports[`Tooltip renders component with correct arrow placement - top 1`] = `
<span
  onClick={[Function]}
>
  <WithStyles(ForwardRef(Tooltip))
    PopperProps={
      Object {
        "popperOptions": Object {
          "modifiers": Object {
            "arrow": Object {
              "element": null,
              "enabled": false,
            },
          },
        },
      }
    }
    classes={
      Object {
        "popper": "arrowPopper-mock",
        "tooltip": "noWrap-mock tooltip-mock",
      }
    }
    interactive={false}
    placement="top"
    title={
      <React.Fragment>
        <span
          className="text-mock"
        >
          Test-title
        </span>
        <span
          className="arrow-mock"
        />
      </React.Fragment>
    }
  >
    <TooltipContent />
  </WithStyles(ForwardRef(Tooltip))>
</span>
`;

exports[`Tooltip renders component with correct arrow placement - top-end 1`] = `
<span
  onClick={[Function]}
>
  <WithStyles(ForwardRef(Tooltip))
    PopperProps={
      Object {
        "popperOptions": Object {
          "modifiers": Object {
            "arrow": Object {
              "element": null,
              "enabled": false,
            },
          },
        },
      }
    }
    classes={
      Object {
        "popper": "arrowPopper-mock",
        "tooltip": "noWrap-mock tooltip-mock",
      }
    }
    interactive={false}
    placement="top-end"
    title={
      <React.Fragment>
        <span
          className="text-mock"
        >
          Test-title
        </span>
        <span
          className="arrow-mock"
        />
      </React.Fragment>
    }
  >
    <TooltipContent />
  </WithStyles(ForwardRef(Tooltip))>
</span>
`;

exports[`Tooltip renders component with correct arrow placement - top-start 1`] = `
<span
  onClick={[Function]}
>
  <WithStyles(ForwardRef(Tooltip))
    PopperProps={
      Object {
        "popperOptions": Object {
          "modifiers": Object {
            "arrow": Object {
              "element": null,
              "enabled": false,
            },
          },
        },
      }
    }
    classes={
      Object {
        "popper": "arrowPopper-mock",
        "tooltip": "noWrap-mock tooltip-mock",
      }
    }
    interactive={false}
    placement="top-start"
    title={
      <React.Fragment>
        <span
          className="text-mock"
        >
          Test-title
        </span>
        <span
          className="arrow-mock"
        />
      </React.Fragment>
    }
  >
    <TooltipContent />
  </WithStyles(ForwardRef(Tooltip))>
</span>
`;

exports[`Tooltip renders component with optional props 1`] = `
<span
  onClick={[Function]}
>
  <WithStyles(ForwardRef(Tooltip))
    PopperProps={
      Object {
        "popperOptions": Object {
          "modifiers": Object {
            "arrow": Object {
              "element": null,
              "enabled": false,
            },
          },
        },
      }
    }
    classes={
      Object {
        "popper": "arrowPopper-mock",
        "tooltip": "noWrap-mock tooltip-mock",
      }
    }
    interactive={true}
    placement="bottom"
    title={
      <React.Fragment>
        <span
          className="text-mock"
        >
          Test-title
        </span>
        <span
          className="arrow-mock"
        />
      </React.Fragment>
    }
  >
    <TooltipContent />
  </WithStyles(ForwardRef(Tooltip))>
</span>
`;
